# Student Loan Risk with Deep Learning

In [38]:
# Imports
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
from pathlib import Path
from tensorflow import keras

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `student-loans.csv` file into a Pandas DataFrame. Review the DataFrame, looking for columns that could eventually define your features and target variables.   

In [2]:
# Read the csv into a Pandas DataFrame
file_path = "https://static.bc-edx.com/ai/ail-v-1-0/m18/lms/datasets/student-loans.csv"
loans_df = pd.read_csv(file_path)

# Review the DataFrame
loans_df.head()

payment_history  location_parameter  stem_degree_score  gpa_ranking  \
0              7.4                0.70               0.00          1.9   
1              7.8                0.88               0.00          2.6   
2              7.8                0.76               0.04          2.3   
3             11.2                0.28               0.56          1.9   
4              7.4                0.70               0.00          1.9   

   alumni_success  study_major_code  time_to_completion  \
0           0.076              11.0                34.0   
1           0.098              25.0                67.0   
2           0.092              15.0                54.0   
3           0.075              17.0                60.0   
4           0.076              11.0                34.0   

   finance_workshop_score  cohort_ranking  total_loan_score  \
0                  0.9978            3.51              0.56   
1                  0.9968            3.20              0.68   
2                  0.9970            3.26              0.65   
3                  0.9980            3.16              0.58   
4                  0.9978            3.51              0.56   

   financial_aid_score  credit_ranking  
0                  9.4               0  
1                  9.8               0  
2                  9.8               0  
3                  9.8               1  
4                  9.4               0

In [3]:
# Review the data types associated with the columns
loans_df.dtypes

payment_history           float64
location_parameter        float64
stem_degree_score         float64
gpa_ranking               float64
alumni_success            float64
study_major_code          float64
time_to_completion        float64
finance_workshop_score    float64
cohort_ranking            float64
total_loan_score          float64
financial_aid_score       float64
credit_ranking              int64
dtype: object

In [4]:
# Check the credit_ranking value counts
loans_df["credit_ranking"].value_counts()

credit_ranking
1    855
0    744
Name: count, dtype: int64

### Step 2: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “credit_ranking”. The remaining columns should define the features dataset.

In [5]:
import numpy as np

In [6]:
# Define the target set y using the credit_ranking column - FIX online help
y = loans_df["credit_ranking"]

# Display a sample of y
y_array = np.array(y)
print(y_array[:5])

[0 0 0 1 0]


In [7]:
# Define features set X by selecting all columns but credit_ranking
X = loans_df.drop('credit_ranking', axis=1)

# Review the  DataFrame
X.head(500)

payment_history  location_parameter  stem_degree_score  gpa_ranking  \
0                7.4                0.70               0.00          1.9   
1                7.8                0.88               0.00          2.6   
2                7.8                0.76               0.04          2.3   
3               11.2                0.28               0.56          1.9   
4                7.4                0.70               0.00          1.9   
..               ...                 ...                ...          ...   
495             10.7                0.35               0.53          2.6   
496              7.8                0.52               0.25          1.9   
497              7.2                0.34               0.32          2.5   
498             10.7                0.35               0.53          2.6   
499              8.7                0.69               0.31          3.0   

     alumni_success  study_major_code  time_to_completion  \
0             0.076              11.0                34.0   
1             0.098              25.0                67.0   
2             0.092              15.0                54.0   
3             0.075              17.0                60.0   
4             0.076              11.0                34.0   
..              ...               ...                 ...   
495           0.070               5.0                16.0   
496           0.081              14.0                38.0   
497           0.090              43.0               113.0   
498           0.070               5.0                16.0   
499           0.086              23.0                81.0   

     finance_workshop_score  cohort_ranking  total_loan_score  \
0                    0.9978            3.51              0.56   
1                    0.9968            3.20              0.68   
2                    0.9970            3.26              0.65   
3                    0.9980            3.16              0.58   
4                    0.9978            3.51              0.56   
..                      ...             ...               ...   
495                  0.9972            3.15              0.65   
496                  0.9984            3.43              0.65   
497                  0.9966            3.32              0.79   
498                  0.9972            3.15              0.65   
499                  1.0002            3.48              0.74   

     financial_aid_score  
0                    9.4  
1                    9.8  
2                    9.8  
3                    9.8  
4                    9.4  
..                   ...  
495                 11.0  
496                  9.0  
497                 11.1  
498                 11.0  
499                 11.6  

[500 rows x 11 columns]

### Step 3: Split the features and target sets into training and testing datasets.


In [8]:
# Split the preprocessed data into a training and testing dataset
# Assign the function a random_state equal to 1
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


### Step 4: Use scikit-learn's `StandardScaler` to scale the features data.

In [9]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
scaler.fit(X_train)

# Fit the scaler to the features training dataset

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)



---

## Compile and Evaluate a Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [10]:
# Define the the number of inputs (features) to the model
num_features = len(X.columns)



# Review the number of features

num_features




11

In [11]:
# Define the number of hidden nodes for the first hidden layer
hn_one = 6

# Define the number of hidden nodes for the second hidden layer
hn_two = 3

# Define the number of neurons in the output layer

nneurons = 1



In [27]:
# Create the Sequential model instance
model = keras.Sequential()

# Add the first hidden layer
model.add(layers.Input(shape=(num_features,)))
model.add(layers.Dense(units=hn_one, activation='relu'))

# model.add(keras.layers.Dense(units=hn_one, activation='relu', input_dim=num_features)) 


# Add the second hidden layer
model.add(layers.Dense(units=hn_two, activation='relu'))
# model.add(keras.layers.Dense(units=hn_two, activation='relu'))

# Add the output layer to the model specifying the number of output neurons and activation function
model.add(layers.Dense(units=nneurons, activation='sigmoid'))
# model.add(keras.layers.Dense(units=nneurons, activation='sigmoid')) 










In [28]:
# Display the Sequential model summary
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                      │ (None, 6)                   │              72 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_7 (Dense)                      │ (None, 3)                   │              21 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_8 (Dense)                      │ (None, 1)                   │               4 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 97 (388.00 B)

 Trainable params: 97 (388.00 B)

 Non-trainable params: 0 (0.00 B)

### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [29]:
# Compile the Sequential model
model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [30]:
# Fit the model using 50 epochs and the training data
fit_model = model.fit(X_train, y_train, epochs=50)

Epoch 1/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5303 - loss: 6.8455
Epoch 2/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5442 - loss: 4.1674
Epoch 3/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5479 - loss: 2.1198
Epoch 4/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5795 - loss: 0.7589
Epoch 5/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6249 - loss: 0.6361
Epoch 6/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5966 - loss: 0.6703
Epoch 7/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6129 - loss: 0.6719
Epoch 8/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6338 - loss: 0.6528
Epoch 9/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6288 - loss: 0.6584
Epoch 10/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6210 - loss: 0.6518
Epoch 11/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6118 - loss: 0.6457
Epoch 12/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6077 - lo

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [31]:
# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
test_loss, test_accuracy = model.evaluate(X_test, y_test)

# Display the model loss and accuracy results
print(f"Test Loss: {test_loss}, Test Accuracy: {test_accuracy}")

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6664 - loss: 0.6223
Test Loss: 0.6188960075378418, Test Accuracy: 0.6524999737739563


### Step 4: Save and export your model to a keras file, and name the file `student_loans.keras`.


In [32]:
# Set the model's file path
file_path = Path("Source_Code/model_model.keras") 

# Export your model to a keras file

model.save(file_path)


---
## Predict Loan Repayment Success by Using your Neural Network Model

### Step 1: Reload your saved model.

In [33]:
# Set the model's file path
file_path = Path("Source_Code/model_model.keras") 

# Load the model to a new object
nn_imported = tf.keras.models.load_model(file_path)

### Step 2: Make predictions on the testing data and save the predictions to a DataFrame.

In [34]:
# Make predictions with the test data
predictions = model.predict(X_test)

# Display a sample of the predictions
predicted_classes = (predictions > 0.5).astype(int) 


print(predictions)


13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step
[[0.58594835]
 [0.32327852]
 [0.6681886 ]
 [0.61135197]
 [0.77076304]
 [0.70606786]
 [0.71211606]
 [0.153249  ]
 [0.6595361 ]
 [0.6001303 ]
 [0.6852342 ]
 [0.52387536]
 [0.47661492]
 [0.6571971 ]
 [0.48739672]
 [0.4910509 ]
 [0.6809204 ]
 [0.62280816]
 [0.6596688 ]
 [0.5351236 ]
 [0.5694202 ]
 [0.5875907 ]
 [0.561857  ]
 [0.6717098 ]
 [0.28996727]
 [0.69983816]
 [0.64335066]
 [0.3745268 ]
 [0.5515122 ]
 [0.0045026 ]
 [0.51885283]
 [0.6984413 ]
 [0.21677282]
 [0.67812216]
 [0.16928874]
 [0.6039454 ]
 [0.5835528 ]
 [0.586334  ]
 [0.7307956 ]
 [0.57194   ]
 [0.6992789 ]
 [0.07560599]
 [0.5383376 ]
 [0.72391796]
 [0.11940268]
 [0.6224879 ]
 [0.4740625 ]
 [0.5884919 ]
 [0.4097612 ]
 [0.5987077 ]
 [0.10139409]
 [0.6534556 ]
 [0.6260817 ]
 [0.06336036]
 [0.40312597]
 [0.65743846]
 [0.609576  ]
 [0.69536066]
 [0.30605474]
 [0.12787399]
 [0.41658634]
 [0.36736074]
 [0.5289099 ]
 [0.3097236 ]
 [0.06002056]
 [0.58816105]
 [0.69753206]
 [0.32343975]
 [0.57148

In [36]:
# Save the predictions to a DataFrame and round the predictions to binary results

predictions_df = pd.DataFrame(columns=["predictions"], data=predictions)

# Round the predictions binar
predictions_df["predictions"] = predictions_df["predictions"].round(0)

# Display 

predictions_df.head(10)



predictions
0          1.0
1          0.0
2          1.0
3          1.0
4          1.0
5          1.0
6          1.0
7          0.0
8          1.0
9          1.0

### Step 4: Display a classification report with the y test data and predictions

In [37]:
# Print the classification report with the y test data and predictions
print(classification_report(y_test, predictions_df))

              precision    recall  f1-score   support

           0       0.69      0.47      0.56       188
           1       0.63      0.82      0.71       212

    accuracy                           0.65       400
   macro avg       0.66      0.64      0.64       400
weighted avg       0.66      0.65      0.64       400



---
## Discuss creating a recommendation system for student loans

Briefly answer the following questions in the space provided:

1. Describe the data that you would need to collect to build a recommendation system to recommend student loan options for students. Explain why this data would be relevant and appropriate. 

2. Based on the data you chose to use in this recommendation system, would your model be using collaborative filtering, content-based filtering, or context-based filtering? Justify why the data you selected would be suitable for your choice of filtering method.

3. Describe two real-world challenges that you would take into consideration while building a recommendation system for student loans. Explain why these challenges would be of concern for a student loan recommendation system.

**1. Describe the data that you would need to collect to build a recommendation system to recommend student loan options for students. Explain why this data would be relevant and appropriate.**  
The model is not very good at prediction defaults and recall worse - 59% vs 78%.  This suggests additonal features - esp regarding credit risk would be helpful to making recommendations - such as current employment, employment history, updated credit history, total amount, maybe adding other external factors such as inflatin rate, unemployment rate - perhaps by location.

**2. Based on the data you chose to use in this recommendation system, would your model be using collaborative filtering, content-based filtering, or context-based filtering? Justify why the data you selected would be suitable for your choice of filtering method.**
Context-based filtering - as new features such as employment changes and credit ratings updated, default risk may also rise and/or fall accordingly.

**3. Describe two real-world challenges that you would take into consideration while building a recommendation system for student loans. Explain why these challenges would be of concern for a student loan recommendation system.**  Data Security - the information on our system is PI and some of it - including if we add credit rating and employment there may be some regulations related to how that type of data is stored/used so we would need to look inot this.  Also, bias - there is a risk that if certain factors such as current employment would improperly weigh against those with lower paying jobs without taking into consideration things such as household expenditures, etc. we would have to ensure that factors are appropreately considered.



